### PR TITLE
Add additional lock timeouts and clean-up `ObjectCacheAppCache`

### PR DESCRIPTION
**Pull request by ronaldbarendse**
This extends PR https://github.com/umbraco/Umbraco-CMS/pull/15902 with additional lock timeouts in `SearchByPredicate()` when entering the upgradable read lock and in `Clear()` when taking the write lock. The timeout exception messages are also updated to clearly state when it occurred, which should make debugging easier.

I've also cleaned up the constructor, as that still had the `IHostEnvironment` parameter (kept in PR https://github.com/umbraco/Umbraco-CMS/pull/15173 to maintain backwards compatibility, but this got merged to v14, so can be breaking now).

### CODE REVIEW

File: src/Umbraco.Core/Cache/ObjectCacheAppCache.cs [hunk 1]
```diff
@@ -34,22 +31,16 @@ public class ObjectCacheAppCache : IAppPolicyCache, IDisposable
     /// Initializes a new instance of the <see cref="ObjectCacheAppCache" />.
     /// </summary>
     public ObjectCacheAppCache()
-        : this(Options.Create(new MemoryCacheOptions()), NullLoggerFactory.Instance, null)
+        : this(new MemoryCacheOptions(), NullLoggerFactory.Instance)
```
Comment (ronaldbarendse) [inline src/Umbraco.Core/Cache/ObjectCacheAppCache.cs:34]: `MemoryCacheOptions` already implements `IOptions<MemoryCacheOptions>`, so there's no reason to use `Options.Create()` 🙌🏻

File: src/Umbraco.Core/Cache/ObjectCacheAppCache.cs [hunk 2]
```diff
@@ -34,22 +31,16 @@ public class ObjectCacheAppCache : IAppPolicyCache, IDisposable
...
     /// <param name="loggerFactory">The logger factory.</param>
-    /// <param name="hostEnvironment">The host environment.</param>
-    public ObjectCacheAppCache(IOptions<MemoryCacheOptions> options, ILoggerFactory loggerFactory, IHostEnvironment? hostEnvironment)
-    {
-        _options = options;
-        _hostEnvironment = hostEnvironment;
-
-        MemoryCache = new MemoryCache(_options, loggerFactory);
-    }
+    public ObjectCacheAppCache(IOptions<MemoryCacheOptions> options, ILoggerFactory loggerFactory)
```
Comment (ronaldbarendse) [inline src/Umbraco.Core/Cache/ObjectCacheAppCache.cs:42]: This constructor is currently not directly used by `CoreScope` and `AppCaches` (or tests), so there's currently no way to customize the options (to set size limits) and/or set the logger factory (to add debug logging). If we don't need this, we could either remove this constructor or make it internal...

Comment (bergmania) [reply src/Umbraco.Core/Cache/ObjectCacheAppCache.cs:42]: Would be okay as internal

Comment (ronaldbarendse) [reply src/Umbraco.Core/Cache/ObjectCacheAppCache.cs:42]:
```suggestion
    internal ObjectCacheAppCache(IOptions<MemoryCacheOptions> options, ILoggerFactory loggerFactory)
```

**Pull request merged**